docs/HighLevelDesignForGraphProgramExtraction.md

```diff
@@ -0,0 +1,70 @@
+3. Avoid the need for aggressive inlining of all TensorFlow-related code on the Swift-side (which is what is currently happening in the Swift implementation).
+4. Allow for easy packaging of Swift for TF libraries that can be used _efficiently_ by client code. By _efficiently_ here we mean that the constructed TF graph will not contain unused code from the imported libraries.
+5. Try to make the _between-communication_ parts of each program as large as possible, as this would allow for more optimizations on either side. We will refer to this as _partitioning_ the two programs into as large chunks as possible, from now on.
+
+This document mainly discusses points 1-4, but also touches upon 5, without going into much detail.
+
+## Tackling Point 1: Single Entrypoints for the Two Programs
+
+Having a single entrypoint for the Swift program is not worth discussing as that is already _handled_ by the fact that, in our design, the Swift program is the one that controls execution anyway, and can invoke TF programs, as needed. In order to be able to have a single entrypoint for the TF program, we first discuss how we construct the graph, such that we have a single TF graph, and then we discuss how the session executing that graph is managed.
+
+### Single TF Graph
+
+We construct the TF graph by _walking through_ the Swift program and creating a program with the exact same structure, but keeping only the TF-relevant parts. This means that we start at the top-level Swift code, that will be executed when we execute the Swift program. For now, we ignore functions as that will be discussed when we tackle point 3.
+
+For all top-level code we check where `Tensor`s are used and how and we built a TF graph, based on the methods outlined in the GPE whitepaper. `Tensor`s that are created on the Swift-side are handled by the TF graph in the way described in the section tackling point 2. After building the TF graph in this manner, we collect all sink nodes in the TF graph, whether those are nodes that send tensors back to the Swift program, or simply nodes that only have a side-effect on the TF side, and we create a new TF `group` node that we call the `sink` node. This means that once the `sink` node finishes executing in TF, all TF-related computation for the current Swift for TF program will have finished executing and so it can be thought of as the TF _endpoint_.
+
+We have ignored functions that use TF for now, but we will discuss those in the section devoted to point 3.
+
+### Single TF Session
+
+Given the single TF graph we have built for the current Swift for TF program, we can then go ahead and add the following in the beggining of the top-level Swift program code (on the Swift side):
+
+1. Create a new TF session (with any desirable configuration).
+2. Add a call to `session.run(sink)`.
```

Comment:
if some sinks are in a control flow (if, while) context, we may not be able to run them in session.run() directly -- but it seems we can make the wrapping if/while nodes as sinks if needed (if their output values have no users).

Reply:
If the TensorFlow `If` and `While` ops produce output tensors, then we can just keep track of where those are used. If they do not, and we rely on side-effects (I'm not sure TF supports that behavior though), then we can consider that `If`/`While` op as a sink node itself and group it along with the other sink nodes. `tf.group` takes a set of ops as input and acts as a no-op barriers that waits for all of them to finish executing, and so should be able to handle this situation.

Reply:
Agree. The functional If and while ops should support "no output tensors" by design, though they are probably not exercised a lot, so some bugs could exist.